README.md

<h1 align="center">
  <p align="center">@qfin/logger</p>
  <p align="center" style="font-size: 0.5em">The most awesome isomorphic logger for NodeJs and Browsers 😎 ❤</p>
</h1><br />
<!-- This file is automatically generated. Please don't edit it directly:
if you find an error, edit the source files, and re-run
./build/generate-docs.js from the project root. -->

<p align="center">
    <a href="https://twitter.com/acdlite/status/974390255393505280"><img src="https://img.shields.io/badge/speed-blazing%20%F0%9F%94%A5-brightgreen.svg" alt="Blazing Fast"></a>
    <a href="https://badge.fury.io/js/%40qfin%2Flogger"><img src="https://badge.fury.io/js/%40qfin%2Flogger.svg" alt="npm version" height="18"></a>
    <a href="https://circleci.com/gh/quantofin/logger"><img src="https://circleci.com/gh/quantofin/logger.svg?style=shield" alt="CircleCI Build Status"></a>
    <a href="https://codecov.io/gh/quantofin/logger">
    <img src="https://codecov.io/gh/quantofin/logger/branch/master/graph/badge.svg" />
    </a>
</p>
<p align="center">
    <a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT"></a>
    <a href="https://snyk.io/test/github/quantofin/logger?targetFile=package.json"><img src="https://snyk.io/test/github/quantofin/logger/badge.svg?targetFile=package.json" alt="Known Vulnerabilities" data-canonical-src="https://snyk.io/test/github/quantofin/logger?targetFile=package.json" style="max-width:100%;"></a>
</p>
<p align="center">
    <a href="https://twitter.com/intent/follow?screen_name=quantofin"><img src="https://img.shields.io/twitter/follow/quantofin.svg?style=social&label=Follow%20@quantofin" alt="Follow on Twitter"></a>  
</p>

# 📦 Features

- Support for [ES6 modules](http://exploringjs.com/es6/ch_modules.html)
- Is a fast JSON logger that supports pretty printing in development mode
- Has static methods, instance methods, child loggers, log file support etc.
- Supports both Browsers and NodeJs

# 💾 Install

**Latest Release:** @qfin/logger:0.0.23

```bash
npm install @qfin/logger
```

# ⚙ Module Documentation

> Documentation is available at [https://quantofin.github.io/logger/](https://quantofin.github.io/logger/)

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

## Logger

The Logger class

Type: [Logger][1]

### Parameters

-   `options` **[object][2]?** Configuration options for the logger.
    -   `options.base` **[object][2]** An object with base properties that will be printed with every log line. (optional, default `null`)
    -   `options.name` **[string][3]** The name of the logger. It is a good practice to give names for easier identification of modules for example. (optional, default `undefined`)
    -   `options.level` **[string][3]** The log level. One of 'fatal', 'error', 'warn', 'info', 'debug', 'trace' or 'silent'. (optional, default `'info'`)
    -   `options.file` **[string][3]** The location of the log file. If not given, logs will be outputted to std out/err. (optional, default `undefined`)
    -   `options.prettyPrint` **[boolean][4]?** Whether to pretty print the logs or output json lines
    -   `options.redact` **([object][2] \| [array][5])?** The object props to redact from the output. As an array, the redact option specifies paths that should have their values redacted from any log output. Each path must be a string using a syntax which corresponds to JavaScript dot and bracket notation. If an object is supplied, three options can be specified: `paths`, `censor` and `remove`.
        -   `options.redact.paths` **[array][5]** Required. An array of paths.
        -   `options.redact.censor` **([string][3] \| [function][6] \| [undefined][7])** Censor option will overwrite keys which are to be redacted. (optional, default `'[Redacted]'`)
        -   `options.redact.remove` **[boolean][4]** Instead of censoring the value, remove both the key and the value. (optional, default `false`)
    -   `options.browser` **[object][2]** Config for browsers only. (optional, default `undefined`)
        -   `options.browser.write` **([function][6] \| [object][2])?** Instead of passing log messages to console.log they can be passed to a supplied function.
            If write is an object, it can have methods that correspond to the levels. When a message is logged at a given level, the corresponding method is called. If a method isn't present, the logging falls back to using the console.

### Examples

```javascript
import Logger from '@qfin/logger';

// this creates a parent logger with name (my-app)
const logger = new Logger({ name: 'my-app', level: 'info' });

logger.info('Hello world from my awesome app!');
logger.debug('This will not be printed');

// --- somewhere else in the codebase

import Logger from '@qfin/logger';

// this creates a child logger with name (db) and it takes the log-level of the parent logger
const dbLogger = new Logger({ name: 'db' });

dbLogger.info('Successfully connected to Database: ', `${db.host}:${db.port}/${db.name}`);

// --- static logger usage

import Logger from '@qfin/logger';

// If logger is already initialized, then the parent logger reference will be used in this static call
// otherwise, a default parent logger will be initialized and then that will be used for logging
Logger.log('This is a static logger log');
```

### log

Write a log with default or set level. The level can be set in the constructor parameter

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### trace

Write a 'trace' level log, if the configured level allows for it.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### debug

Write a 'debug' level log, if the configured level allows for it.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### info

Write a 'info' level log, if the configured level allows for it.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### warn

Write a 'warn' level log, if the configured level allows for it.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### error

Write a 'error' level log, if the configured level allows for it.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### fatal

Write a 'fatal' level log, if the configured level allows for it.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### log

Static Method

Write a log with default or set level.

If the logger has already been initialized then the log level will be referred from the initialized logger,
otherwise a new logger will be initialized and its default properties will be set and used.

#### Parameters

-   `arg0` **[object][2]?** An object can optionally be supplied as the first parameter. Each enumerable key and value of the mergingObject is copied in to the JSON log line.
-   `args` **...any?** `args` are of type varargs and each of the args can take various forms as follows:-   A message string can optionally be supplied as the first parameter, or as the second parameter after supplying a mergingObject.
        By default, the contents of the message parameter will be merged into the JSON log line under the msg key.

    -   All arguments supplied after message are serialized and interpolated according to any supplied printf-style placeholders (%s, %d, %o|%O|%j) or else concatenated together with the message string to form the final output msg value for the JSON log line.

#### Examples

```javascript
logger.info({MIX: {IN: true}})
// {"level":30,"time":1531254555820,"MIX":{"IN":true},"v":1}

logger.info('hello world')
// {"level":30,"time":1531257112193,"msg":"hello world","v":1}

logger.info('hello', 'world')
// {"level":30,"time":1531257618044,"msg":"hello world","v":1}

logger.info('hello', {worldly: 1})
// {"level":30,"time":1531257797727,"msg":"hello {\"worldly\":1}","v":1}

logger.info('%o hello', {worldly: 1})
// {"level":30,"time":1531257826880,"msg":"{\"worldly\":1} hello","v":1}
```

### trace

Static Method

Write a 'trace' level log, if the configured level allows for it.

If the logger has already been initialized then the log level will be referred from the initialized logger,
otherwise a new logger will be initialized and its default properties will be set and used.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### debug

Static Method

Write a 'debug' level log, if the configured level allows for it.

If the logger has already been initialized then the log level will be referred from the initialized logger,
otherwise a new logger will be initialized and its default properties will be set and used.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### info

Static Method

Write a 'info' level log, if the configured level allows for it.

If the logger has already been initialized then the log level will be referred from the initialized logger,
otherwise a new logger will be initialized and its default properties will be set and used.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### warn

Static Method

Write a 'warn' level log, if the configured level allows for it.

If the logger has already been initialized then the log level will be referred from the initialized logger,
otherwise a new logger will be initialized and its default properties will be set and used.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### error

Static Method

Write a 'error' level log, if the configured level allows for it.

If the logger has already been initialized then the log level will be referred from the initialized logger,
otherwise a new logger will be initialized and its default properties will be set and used.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

### fatal

Static Method

Write a 'fatal' level log, if the configured level allows for it.

If the logger has already been initialized then the log level will be referred from the initialized logger,
otherwise a new logger will be initialized and its default properties will be set and used.

#### Parameters

-   `arg0` **[object][2]?** See [Logger's log method for more info][Logger.log][8].
-   `args` **...any?** See [Logger's log method for more info][Logger.log][8].

[1]: #logger

[2]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object

[3]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String

[4]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean

[5]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array

[6]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function

[7]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/undefined

[8]: #loggerlog


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [Unreleased](https://github.com/quantofin/logger/compare/v0.0.22...HEAD)

### Merged

- Bump eslint-config-prettier from 4.3.0 to 5.0.0 [`#158`](https://github.com/quantofin/logger/pull/158)
- Bump prettier from 1.17.1 to 1.18.2 [`#154`](https://github.com/quantofin/logger/pull/154)
- Bump prettier from 1.17.0 to 1.17.1 [`#128`](https://github.com/quantofin/logger/pull/128)
- Bump documentation from 10.1.0 to 11.0.0 [`#123`](https://github.com/quantofin/logger/pull/123)
- Bump release-it from 11.0.3 to 12.0.0 [`#116`](https://github.com/quantofin/logger/pull/116)
- Bump release-it from 10.4.5 to 11.0.2 [`#112`](https://github.com/quantofin/logger/pull/112)
- Bump dotenv from 7.0.0 to 8.0.0 [`#113`](https://github.com/quantofin/logger/pull/113)
- Bump eslint-plugin-node from 8.0.1 to 9.0.0 [`#114`](https://github.com/quantofin/logger/pull/114)

### Commits

- update packages [`bae27b4`](https://github.com/quantofin/logger/commit/bae27b4519860ade09de0269fb17b5f4f160ab0f)
- Bump jest from 24.7.1 to 24.8.0 [`cd63064`](https://github.com/quantofin/logger/commit/cd6306472c9ecd8c121c361b032a7f280738e624)
- Bump documentation from 11.0.0 to 11.0.1 [`37bbcb0`](https://github.com/quantofin/logger/commit/37bbcb0b0a8de8dc50963bb17d1d88360e940739)
- Bump terser-webpack-plugin from 1.2.3 to 1.2.4 [`f8ac968`](https://github.com/quantofin/logger/commit/f8ac968f9e0d5094f840fac6490821ff9943298e)
- Bump webpack-cli from 3.3.4 to 3.3.5 [`63ccac9`](https://github.com/quantofin/logger/commit/63ccac97dc7eada656070ce730774f86cff0600a)
- Bump babel-jest from 24.7.1 to 24.8.0 [`f01bf55`](https://github.com/quantofin/logger/commit/f01bf5526ed6b724049444ad2c3b56a19caf614d)
- Bump @babel/preset-env from 7.4.4 to 7.4.5 [`604a87d`](https://github.com/quantofin/logger/commit/604a87d4f8fa990e9a2538b6d3ba2d73020c872b)
- Bump release-it from 12.0.1 to 12.1.0 [`ad10288`](https://github.com/quantofin/logger/commit/ad10288e1550a17c8f82569864617d1e45f308dc)
- Bump babel-loader from 8.0.5 to 8.0.6 [`5dcf748`](https://github.com/quantofin/logger/commit/5dcf748310bed6db52a4355f57d158113d496de4)
- Bump terser-webpack-plugin from 1.2.4 to 1.3.0 [`85f6da5`](https://github.com/quantofin/logger/commit/85f6da5fe54b6bbaac4f38e9b61d2adef4006555)
- Bump @babel/core from 7.4.4 to 7.4.5 [`5a30327`](https://github.com/quantofin/logger/commit/5a3032705222dad943f485d8ab3010c83aa0b11d)
- Bump webpack from 4.31.0 to 4.32.0 [`48a42b8`](https://github.com/quantofin/logger/commit/48a42b8926d0fcdb23c155b71fdce67881eee1e1)
- Bump release-it from 12.2.2 to 12.3.0 [`013e148`](https://github.com/quantofin/logger/commit/013e1485878d83d8642e1175633b34a9600dbc00)
- Bump pino-pretty from 3.0.0 to 3.0.1 [`147cae1`](https://github.com/quantofin/logger/commit/147cae188f97229b90fe8a874d482032a69d4800)
- Bump eslint-plugin-node from 9.0.1 to 9.1.0 [`da5270c`](https://github.com/quantofin/logger/commit/da5270cd2fbc3ff619502f0b1882b37b96d9804b)
- Bump pino-pretty from 3.1.0 to 3.2.0 [`4d0d2f1`](https://github.com/quantofin/logger/commit/4d0d2f14ca78f234614148722ef900bbe9f47cd8)
- Bump eslint-plugin-import from 2.17.2 to 2.17.3 [`0df2429`](https://github.com/quantofin/logger/commit/0df2429c8cdbaab0ac6344a81b6fb71bbd742ba7)
- Bump webpack-cli from 3.3.2 to 3.3.3 [`4a49dea`](https://github.com/quantofin/logger/commit/4a49dea29b625bdcaf1688640284312fcf19e8df)
- Bump eslint-plugin-import from 2.17.3 to 2.18.0 [`a352ff3`](https://github.com/quantofin/logger/commit/a352ff394a69348f21a232873307a0d38862192a)
- Bump release-it from 12.2.0 to 12.2.1 [`2353119`](https://github.com/quantofin/logger/commit/23531195ec9928afe916a90b64f71100f475ee23)
- Bump pino from 5.12.3 to 5.12.4 [`34e70f9`](https://github.com/quantofin/logger/commit/34e70f9353babe80cb20cf68c6c6d2155e8482e8)
- Bump pino from 5.12.5 to 5.12.6 [`c4056c2`](https://github.com/quantofin/logger/commit/c4056c2af002017e8f6034b4ac3da212f175f7d2)
- Bump release-it from 12.2.1 to 12.2.2 [`8610d81`](https://github.com/quantofin/logger/commit/8610d8198cb3da91631023135ab6554761377a9d)
- Bump webpack from 4.32.2 to 4.33.0 [`3c59aa1`](https://github.com/quantofin/logger/commit/3c59aa1ed9db0ddd07abc1ed797660fc7fb7eeb4)
- Bump release-it from 12.0.0 to 12.0.1 [`2963135`](https://github.com/quantofin/logger/commit/29631358918dcd0c12237fe7817df4dd1953bb1c)
- Bump pino from 5.12.4 to 5.12.5 [`594acdf`](https://github.com/quantofin/logger/commit/594acdfcca0f290ad663bee93420f48fac1c6e5b)
- Bump codecov from 3.3.0 to 3.4.0 [`ff957b4`](https://github.com/quantofin/logger/commit/ff957b400f4c723ebd7f437601f2a4ce0aa275dc)
- Bump codecov from 3.4.0 to 3.5.0 [`e0c374a`](https://github.com/quantofin/logger/commit/e0c374aabf6f4b31ce684fa855401d64397d9f66)
- Merge pull request #147 from quantofin/dependabot/npm_and_yarn/develop/release-it-12.2.1 [`bb0e30f`](https://github.com/quantofin/logger/commit/bb0e30f6f769fc8d0b84ce9adf11cc80663d447d)
- Bump pino-pretty from 3.0.1 to 3.1.0 [`3c96650`](https://github.com/quantofin/logger/commit/3c96650c973859265875332f6374f867cc016044)
- Merge pull request #146 from quantofin/dependabot/npm_and_yarn/develop/eslint-plugin-import-2.17.3 [`4876ff2`](https://github.com/quantofin/logger/commit/4876ff2115fef61f8102ad60abd460f6ee49c786)
- Merge pull request #160 from quantofin/dependabot/npm_and_yarn/develop/pino-pretty-3.2.0 [`71748a2`](https://github.com/quantofin/logger/commit/71748a2927cce1c6db7a082705006d1e179b45eb)
- Merge pull request #145 from quantofin/dependabot/npm_and_yarn/develop/terser-webpack-plugin-1.3.0 [`aa582e9`](https://github.com/quantofin/logger/commit/aa582e9a698600cf1dd7c7a948dfabc9b726456a)

src/modules/__tests__/.eslintrc.js
module.exports = {
    env: {
      jest: true,
    },
};
    

src/modules/__tests__/logger.spec.mjs
import Logger from '../logger';

describe('Logger module', () => {
  it('should initialize a Logger object', () => {
    expect(new Logger()).toBeInstanceOf(Logger);
  });

  it('should have a log object with logging methods', () => {
    const logger = new Logger();
    expect(logger.log).toBeDefined();
    expect(typeof logger.log).toEqual('function');
    expect(logger.fatal).toBeDefined();
    expect(typeof logger.fatal).toEqual('function');
    expect(logger.error).toBeDefined();
    expect(typeof logger.error).toEqual('function');
    expect(logger.warn).toBeDefined();
    expect(typeof logger.warn).toEqual('function');
    expect(logger.info).toBeDefined();
    expect(typeof logger.info).toEqual('function');
    expect(logger.debug).toBeDefined();
    expect(typeof logger.debug).toEqual('function');
    expect(logger.trace).toBeDefined();
    expect(typeof logger.trace).toEqual('function');
  });
});


.eslintrc.js
// http://eslint.org/docs/user-guide/configuring

module.exports = {
  root: true,
  parser: 'babel-eslint',
  parserOptions: {
    sourceType: 'module',
  },
  env: {
    browser: true,
  },
  extends: ['airbnb-base', 'plugin:prettier/recommended'],
  plugins: ['prettier'],
  // check if imports actually resolve
  settings: {
    'import/resolver': {
      webpack: {
        config: 'webpack.config.js',
      },
    },
  },
  // add your custom rules here
  rules: {
    // don't require .js extension when importing
    'import/extensions': [
      'error',
      'always',
      {
        js: 'never',
        jsx: 'never',
        mjs: 'never',
      },
    ],
    // allow optionalDependencies
    'import/no-extraneous-dependencies': [
      'error',
      {
        optionalDependencies: ['test/unit/index.js'],
      },
    ],
    // allow debugger during development
    'no-debugger': process.env.NODE_ENV === 'production' ? 2 : 0,

    'object-curly-newline': [
      'error',
      {
        ObjectExpression: { multiline: true, minProperties: 6 },
        ObjectPattern: { multiline: true, minProperties: 6 },
        ImportDeclaration: 'never',
        ExportDeclaration: { multiline: true, minProperties: 6 },
      },
    ],
    'prettier/prettier': 'error',
  },
};


jest.config.js
const fs = require('fs');
const path = require('path');

const ENV = process.env.NODE_ENV || 'test';
const isTest = ENV === 'test';

// --- Ignore paths... update this as needed
const ignorePatterns = ['/node_modules/', '/dist/', '/build/', '/coverage/', '/docs/', '/examples/'];

let mod = fs
  .readdirSync(path.join(__dirname, 'src'))
  .filter((p) => {
    const stats = fs.statSync(path.join(__dirname, 'src', p));
    return stats.isDirectory() && !['__tests__'].includes(p);
  })
  .reduce((acc, val) => `${acc}${val}|`, '');

if (mod.endsWith('|')) mod = mod.substring(0, mod.length - 1);

mod = `^@(${mod})(.*)$`;

module.exports = {
  verbose: isTest,
  moduleFileExtensions: ['js', 'mjs', 'json', 'node'],
  modulePathIgnorePatterns: ignorePatterns,
  moduleNameMapper: { [mod]: '<rootDir>/src/$1$2' },
  transform: { '^.+\\.m?jsx?$': 'babel-jest' },
  testMatch: ['**/__tests__/*.spec.(js|mjs|ts)'],
  collectCoverage: isTest,
  collectCoverageFrom: ['<rootDir>/src/**/*.{js,mjs}', ...ignorePatterns.map((p) => `!**${p}**`)],
  coveragePathIgnorePatterns: ignorePatterns,
};


src/modules/logger.mjs
/* eslint-disable no-param-reassign */
import pino from 'pino';

const superstruct = require('superstruct');

function addFinalHandler(logger) {
  // use pino.final to create a special logger that
  // guarantees final tick writes
  const handler = pino.final(logger, (err, finalLogger, evt) => {
    finalLogger.info(`${evt} caught`);
    if (err) finalLogger.error(err, 'error caused exit');
    process.exit(err ? 1 : 0);
  });
  // catch all the ways node might exit
  process.on('uncaughtException', (err) => handler(err, 'uncaughtException'));
  process.on('unhandledRejection', (err) => handler(err, 'unhandledRejection'));
  process.on('SIGINT', () => handler(null, 'SIGINT'));
  process.on('SIGQUIT', () => handler(null, 'SIGQUIT'));
  process.on('SIGTERM', () => handler(null, 'SIGTERM'));
}

export default class Logger {
  static LoggerProps = superstruct.struct(
    {
      prettyPrint: 'boolean?',
      name: 'string?',
      level: 'string?',
      base: 'object|null?',
      redact: 'object|array?',
      file: 'string?',
      browser: 'object?',
    },
    {
      prettyPrint: process.env.NODE_ENV === 'development',
      base: null,
      name: undefined,
      level: 'info',
      file: undefined,
      browser: undefined,
      redact: ['secret', 'username', 'password', 'email', 'phone'],
    }
  );

  static logger;

  static logLevel;

  /**
   * The Logger class
   * @public
   * @module @sa/logger
   * @class Logger
   * @namespace Logger
   * @type {Logger}
   *
   * Creates either a new logger instance or a child instance internally based on [Pino]{@link https://getpino.io}
   *
   * The first time you instantiate the Logger, it will create a new instance and every other call will instantiate a child logger.
   *
   * @param {object} [options] - Configuration options for the logger.
   * @param {object} [options.base=null] - An object with base properties that will be printed with every log line.
   * @param {string} [options.name=undefined] - The name of the logger. It is a good practice to give names for easier identification of modules for example.
   * @param {string} [options.level='info'] - The log level. One of 'fatal', 'error', 'warn', 'info', 'debug', 'trace' or 'silent'.
   * @param {string} [options.file=undefined] - The location of the log file. If not given, logs will be outputted to std out/err.
   * @param {boolean} [options.prettyPrint] - Whether to pretty print the logs or output json lines
   * @param {object|array} [options.redact] - The object props to redact from the output. As an array, the redact option specifies paths that should have their values redacted from any log output. Each path must be a string using a syntax which corresponds to JavaScript dot and bracket notation. If an object is supplied, three options can be specified: `paths`, `censor` and `remove`.
   * @param {array} options.redact.paths - Required. An array of paths.
   * @param {string|function|undefined} [options.redact.censor='[Redacted]'] - Censor option will overwrite keys which are to be redacted.
   * @param {boolean} [options.redact.remove=false] - Instead of censoring the value, remove both the key and the value.
   * @param {object} [options.browser=undefined] - Config for browsers only.
   * @param {function|object} [options.browser.write] - Instead of passing log messages to console.log they can be passed to a supplied function.
   * If write is an object, it can have methods that correspond to the levels. When a message is logged at a given level, the corresponding method is called. If a method isn't present, the logging falls back to using the console.
   *
   * @example
   *
   * import Logger from '@qfin/logger';
   *
   * // this creates a parent logger with name (my-app)
   * const logger = new Logger({ name: 'my-app', level: 'info' });
   *
   * logger.info('Hello world from my awesome app!');
   * logger.debug('This will not be printed');
   *
   * // --- somewhere else in the codebase
   *
   * import Logger from '@qfin/logger';
   *
   * // this creates a child logger with name (db) and it takes the log-level of the parent logger
   * const dbLogger = new Logger({ name: 'db' });
   *
   * dbLogger.info('Successfully connected to Database: ', `${db.host}:${db.port}/${db.name}`);
   *
   * // --- static logger usage
   *
   * import Logger from '@qfin/logger';
   *
   * // If logger is already initialized, then the parent logger reference will be used in this static call
   * // otherwise, a default parent logger will be initialized and then that will be used for logging
   * Logger.log('This is a static logger log');
   *
   */
  constructor(options = {}) {
    // if parent logger is available and log level is not passed, then inherit the parent's log level
    if (!options.level && Logger.logger) options.level = Logger.logger.level;

    const props = Logger.LoggerProps(options);

    if (Logger.logger) {
      this.loggerInstance = Logger.logger.child({
        name: props.name,
        level: props.level,
      });
    } else {
      this.loggerInstance = Logger.instantiateLogger(options);
      addFinalHandler(this.loggerInstance); // only add handlers once
    }
    this.logLevel = props.level;
  }

  static instantiateLogger(options = {}) {
    const props = Logger.LoggerProps(options);
    const lProps = {
      ...props,
      file: undefined,
      prettyPrint: props.prettyPrint,
    };

    if (props.prettyPrint) {
      lProps.timestamp = () => JSON.stringify({ time: new Date().toLocaleString() });
    }
    Logger.logger = pino(lProps, pino.destination(props.file));
    Logger.logLevel = props.level;

    Logger.logger[Logger.logLevel](`Logger is initialized with ${props.level} level`);
    return Logger.logger;
  }

  /**
   * Static Method
   *
   * Write a log with default or set level.
   *
   * If the logger has already been initialized then the log level will be referred from the initialized logger,
   * otherwise a new logger will be initialized and its default properties will be set and used.
   *
   * @static
   * @public
   *
   * @param {object} [arg0] - An object can optionally be supplied as the first parameter. Each enumerable key and value of the mergingObject is copied in to the JSON log line.
   *
   * @param {...any} [args]
   *
   * `args` are of type varargs and each of the args can take various forms as follows:
   *
   * - A message string can optionally be supplied as the first parameter, or as the second parameter after supplying a mergingObject.
   *   By default, the contents of the message parameter will be merged into the JSON log line under the msg key.
   *
   * - All arguments supplied after message are serialized and interpolated according to any supplied printf-style placeholders (%s, %d, %o|%O|%j) or else concatenated together with the message string to form the final output msg value for the JSON log line.
   *
   * @example
   *
   * logger.info({MIX: {IN: true}})
   * // {"level":30,"time":1531254555820,"MIX":{"IN":true},"v":1}
   *
   * logger.info('hello world')
   * // {"level":30,"time":1531257112193,"msg":"hello world","v":1}
   *
   * logger.info('hello', 'world')
   * // {"level":30,"time":1531257618044,"msg":"hello world","v":1}
   *
   * logger.info('hello', {worldly: 1})
   * // {"level":30,"time":1531257797727,"msg":"hello {\"worldly\":1}","v":1}
   *
   * logger.info('%o hello', {worldly: 1})
   * // {"level":30,"time":1531257826880,"msg":"{\"worldly\":1} hello","v":1}
   *
   */
  static log(arg0, ...args) {
    if (!Logger.logger) {
      Logger.instantiateLogger();
    }
    return Logger.logger[Logger.logLevel](arg0, ...args);
  }

  /**
   *  Write a log with default or set level. The level can be set in the constructor parameter
   *
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  log(arg0, ...args) {
    return this.loggerInstance[this.logLevel](arg0, ...args);
  }

  /**
   * Static Method
   *
   * Write a 'trace' level log, if the configured level allows for it.
   *
   * If the logger has already been initialized then the log level will be referred from the initialized logger,
   * otherwise a new logger will be initialized and its default properties will be set and used.
   *
   * @static
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  static trace(arg0, ...args) {
    if (!Logger.logger) {
      Logger.instantiateLogger();
    }
    return Logger.logger.trace(arg0, ...args);
  }

  /**
   * Write a 'trace' level log, if the configured level allows for it.
   *
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  trace(arg0, ...args) {
    return this.loggerInstance.trace(arg0, ...args);
  }

  /**
   * Static Method
   *
   * Write a 'debug' level log, if the configured level allows for it.
   *
   * If the logger has already been initialized then the log level will be referred from the initialized logger,
   * otherwise a new logger will be initialized and its default properties will be set and used.
   *
   * @static
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  static debug(arg0, ...args) {
    if (!Logger.logger) {
      Logger.instantiateLogger();
    }
    return Logger.logger.debug(arg0, ...args);
  }

  /**
   * Write a 'debug' level log, if the configured level allows for it.
   *
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  debug(arg0, ...args) {
    return this.loggerInstance.debug(arg0, ...args);
  }

  /**
   * Static Method
   *
   * Write a 'info' level log, if the configured level allows for it.
   *
   * If the logger has already been initialized then the log level will be referred from the initialized logger,
   * otherwise a new logger will be initialized and its default properties will be set and used.
   *
   * @static
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  static info(arg0, ...args) {
    if (!Logger.logger) {
      Logger.instantiateLogger();
    }
    return Logger.logger.info(arg0, ...args);
  }

  /**
   * Write a 'info' level log, if the configured level allows for it.
   *
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  info(arg0, ...args) {
    return this.loggerInstance.info(arg0, ...args);
  }

  /**
   * Static Method
   *
   * Write a 'warn' level log, if the configured level allows for it.
   *
   * If the logger has already been initialized then the log level will be referred from the initialized logger,
   * otherwise a new logger will be initialized and its default properties will be set and used.
   *
   * @static
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  static warn(arg0, ...args) {
    if (!Logger.logger) {
      Logger.instantiateLogger();
    }
    return Logger.logger.warn(arg0, ...args);
  }

  /**
   * Write a 'warn' level log, if the configured level allows for it.
   *
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  warn(arg0, ...args) {
    return this.loggerInstance.warn(arg0, ...args);
  }

  /**
   * Static Method
   *
   * Write a 'error' level log, if the configured level allows for it.
   *
   * If the logger has already been initialized then the log level will be referred from the initialized logger,
   * otherwise a new logger will be initialized and its default properties will be set and used.
   *
   * @static
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  static error(arg0, ...args) {
    if (!Logger.logger) {
      Logger.instantiateLogger();
    }
    return Logger.logger.error(arg0, ...args);
  }

  /**
   * Write a 'error' level log, if the configured level allows for it.
   *
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  error(arg0, ...args) {
    return this.loggerInstance.error(arg0, ...args);
  }

  /**
   * Static Method
   *
   * Write a 'fatal' level log, if the configured level allows for it.
   *
   * If the logger has already been initialized then the log level will be referred from the initialized logger,
   * otherwise a new logger will be initialized and its default properties will be set and used.
   *
   * @static
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  static fatal(arg0, ...args) {
    if (!Logger.logger) {
      Logger.instantiateLogger();
    }
    return Logger.logger.fatal(arg0, ...args);
  }

  /**
   * Write a 'fatal' level log, if the configured level allows for it.
   *
   * @public
   * @param {object} [arg0] - See [Logger's log method for more info]{@link Logger.log}.
   * @param {...any} [args] - See [Logger's log method for more info]{@link Logger.log}.
   */
  fatal(arg0, ...args) {
    return this.loggerInstance.fatal(arg0, ...args);
  }
}
